usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -0,0 +1,291 @@
+
+            for field in self.fields:
+                row[field] = record.get(self.type_code_to_field_map[record[self.award_semaphore]].get(field))
+
+            if "Award ID" in self.fields:
+                for id_type in self.award_id_fields:
+                    if record[id_type]:
+                        row["Award ID"] = record[id_type]
+                        break
+
+            results.append(row)
+
+        self.populate_response(results=results, has_next=has_next)
+
+    def _construct_award_variables(self, filters):
+        """ Award logic for SpendingByAward"""
+
+        # build sql query filters
+        model = obtain_view_from_award_group(filters["award_type_codes"])
+        self.queryset = matview_search_filter(filters, model).values()
+        self.award_semaphore = "type"
+        minimum_db_fields = {"award_id", "piid", "fain", "uri", "type"}
+        self.award_id_fields = ["piid", "fain", "uri"]
+        self.internal_id_field = "award_id"
+        api_to_db_mapping_list = [
+            award_contracts_mapping,
+            loan_award_mapping,
+            non_loan_assistance_award_mapping,
+            award_idv_mapping,
+        ]
+
+        self.database_fields = self.get_database_fields(minimum_db_fields, api_to_db_mapping_list)
+
+        # Modify queryset to be ordered by requested "sort" in the request or default value(s)
+        if self.pagination["sort_key"] == "Award ID":
+            self.sort_by_fields = self.award_id_fields
+        else:
+            if set(filters["award_type_codes"]) <= set(contract_type_mapping):
+                self.sort_by_fields = [award_contracts_mapping[self.pagination["sort_key"]]]
+            elif set(filters["award_type_codes"]) <= set(loan_type_mapping):
+                self.sort_by_fields = [loan_award_mapping[self.pagination["sort_key"]]]
+            elif set(filters["award_type_codes"]) <= set(idv_type_mapping):
+                self.sort_by_fields = [award_idv_mapping[self.pagination["sort_key"]]]
+            elif set(filters["award_type_codes"]) <= set(non_loan_assistance_type_mapping):
+                self.sort_by_fields = [non_loan_assistance_award_mapping[self.pagination["sort_key"]]]
+            else:
+                msg = "Sort key '{}' doesn't exist for the selected Award group".format(self.pagination["sort_key"])
+                raise InvalidParameterException(msg)
+
+        self.type_code_to_field_map = {}
+        for award_type in loan_type_mapping:
+            self.type_code_to_field_map[award_type] = loan_award_mapping
+        for award_type in non_loan_assistance_type_mapping:
+            self.type_code_to_field_map[award_type] = non_loan_assistance_award_mapping
+        for award_type in award_idv_mapping:
+            self.type_code_to_field_map[award_type] = award_idv_mapping
+        for award_type in contract_type_mapping:
+            self.type_code_to_field_map[award_type] = award_contracts_mapping
```

Comment:
Will more than one of these ever apply? If not, why not use some kind of enum to indicate which kind of award is being processed, and use that to sort which mapping to apply. Similar usages might be able to shorten this code by a lot, and increase adaptability if and when this splitting strategy changes.

Reply:
way ahead of you :)